src/solvers/dg/dg.jl

```diff
@@ -54,9 +54,399 @@ end
+  Entropy stable high order discontinuous Galerkin methods with suitable
+  quadrature rules for hyperbolic conservation laws
+  [doi: 10.1016/j.jcp.2017.05.025](https://doi.org/10.1016/j.jcp.2017.05.025)
+"""
+struct VolumeIntegralFluxDifferencing{VolumeFlux} <: AbstractVolumeIntegral
+  volume_flux::VolumeFlux
+end
+
+"""
+    VolumeIntegralShockCapturingHG
+
+Shock-capturing volume integral type for DG methods proposed by
+- Hennemann, Gassner (2020)
+  "A provably entropy stable subcell shock capturing approach for high order split form DG"
+  [arXiv: 2008.12044](https://arxiv.org/abs/2008.12044)
+"""
+struct VolumeIntegralShockCapturingHG{VolumeFluxDG, VolumeFluxFV, Indicator} <: AbstractVolumeIntegral
+  volume_flux_dg::VolumeFluxDG # symmetric, e.g. split-form or entropy-conservative
+  volume_flux_fv::VolumeFluxFV # non-symmetric in general, e.g. entropy-dissipative
+  indicator::Indicator
+end
+
+function VolumeIntegralShockCapturingHG(indicator; volume_flux_dg=flux_central,
+                                                   volume_flux_fv=flux_lax_friedrichs)
+  VolumeIntegralShockCapturingHG{typeof(volume_flux_dg), typeof(volume_flux_fv), typeof(indicator)}(
+    volume_flux_dg, volume_flux_fv, indicator)
+end
+
+
+function get_element_variables!(element_variables, u, mesh, equations,
+                                volume_integral::VolumeIntegralShockCapturingHG, dg, cache)
+  # call the indicator to get up-to-date values for IO
+  volume_integral.indicator(u, equations, dg, cache)
+  get_element_variables!(element_variables, volume_integral.indicator, volume_integral)
+end
+
+
+
+abstract type AbstractBasisSBP{RealT<:Real} end
+
+abstract type AbstractMortar{RealT<:Real} end
+
+abstract type MortarL2{RealT<:Real} <: AbstractMortar{RealT} end
+
+
+struct DG{RealT, Basis<:AbstractBasisSBP{RealT}, Mortar, SurfaceFlux, VolumeIntegral}
+  basis::Basis
+  mortar::Mortar
+  surface_flux::SurfaceFlux
+  volume_integral::VolumeIntegral
+end
+
+function Base.show(io::IO, dg::DG{RealT}) where {RealT}
+  print(io, "DG{", RealT, "}(")
+  print(io,       dg.basis)
+  print(io, ", ", dg.mortar)
+  print(io, ", ", dg.surface_flux)
+  print(io, ", ", dg.volume_integral)
+  print(io, ")")
+end
+
+function Base.show(io::IO, ::MIME"text/plain", dg::DG{RealT}) where {RealT}
+  println(io, "DG{", RealT, "} using")
+  println(io, "- ", dg.basis)
+  println(io, "- ", dg.mortar)
+  println(io, "- ", dg.surface_flux)
+  print(io,   "- ", dg.volume_integral)
+end
+
+@inline Base.real(dg::DG{RealT}) where {RealT} = RealT
+
+# TODO: Taal refactor, use case?
+# Deprecate in favor of nnodes or order_of_accuracy?
+@inline polydeg(dg::DG) = polydeg(dg.basis)
+
+@inline ndofs(mesh::TreeMesh, dg::DG, cache) = nelements(cache.elements) * nnodes(dg)^ndims(mesh)
+
+
+function get_element_variables!(element_variables, u, mesh, equations, dg::DG, cache)
+  get_element_variables!(element_variables, u, mesh, equations, dg.volume_integral, dg, cache)
+end
```

Comment:
Because I've sorted this file by object and not by verb. For example, it doesn't make sense to me to collect all `show` methods in one place. Instead, it feels more natural to place them near to the objects which are `show`n.

Reply:
Gotcha! Thanks for the explanation.